backend/OKPluginVariableRendering/VariableRendering.cs

```diff
@@ -0,0 +1,342 @@
+            {
+                var layerRelations = await relationModel.GetRelations(RelationSelectionAll.Instance, layer, trans, changesetProxy.TimeThreshold);
+
+                allRelations.AddRange(layerRelations);
+            }
+
+            var fragments = new List<BulkCIAttributeDataLayerScope.Fragment>();
+
+            foreach (var mainCI in mainCIs)
+            {
+
+                var gatheredAttributes = new Dictionary<string, GatheredAttribute>();
+
+                foreach (var mapping in cfg.BaseCI.AttributeMapping)
+                {
+                    foreach (var (name, attribute) in mainCI.MergedAttributes)
+                    {
+                        if (IsAttributeIncludedInSource(name, mapping.Source))
+                        {
+                            var a = new GatheredAttribute
+                            {
+                                Name = GetTargetName(name, mapping.Target),
+                                Value = attribute.Attribute.Value,
+                                Priority = baseCIPrio,
+                            };
+
+                            if (gatheredAttributes.ContainsKey(a.Name))
+                            {
+                                // check priority of attributes
+                                if (gatheredAttributes[a.Name].Priority < a.Priority)
+                                {
+                                    gatheredAttributes[a.Name] = a;
+                                }
+                            }
+                            else
+                            {
+                                gatheredAttributes.Add(a.Name, a);
+                            }
+                        }
+                    }
+                }
+
+
+                foreach (var followRelation in cfg.BaseCI.FollowRelations)
+                {
+                    var prevCI = mainCI;
+                    foreach (var follow in followRelation.Follow)
+                    {
+                        // get the predicate remove the first char, first char defines the direction of relation
+                        var predicate = follow.Predicate[1..];
+
+                        // check relation direction
+                        var r = allRelations.Where(r =>
+                        {
+                            if (follow.Predicate[0] == '>')
+                            {
+                                return r.FromCIID == prevCI.ID && r.PredicateID == predicate;
+                            }
+                            else
+                            {
+                                return r.ToCIID == prevCI.ID && r.PredicateID == predicate;
+                            }
+                        }).FirstOrDefault();
+
+
+                        if (r == null)
+                        {
+                            break;
+                        }
+
+                        var targetCI = allCIs.Where(ci =>
```

Comment:
The same applies here as above for the relations... instead of doing a O(N) operation for each mainCI, pre-create a dictionary with CIID as key for a O(1) lookup.

If you do these two things, your implementation should be much faster imho.